src/AI/Pathfinder.cpp

```diff
@@ -0,0 +1,448 @@
+    }
+
+    float Pathfinder::calculateSlopeFromNormal(const bs::Vector3& normal) const
+    {
+      return acos(normal.y);
+    }
+
+    HWaypoint Pathfinder::findNextVisibleWaypoint(const bs::Vector3& from) const
+    {
+      // TODO: Check for obstructions
+      return mWaynet->findClosestWaypointTo(from);
+    }
+
+    bool Pathfinder::hasActiveRouteBeenCompleted(const bs::Vector3& positionNow) const
+    {
+      if (!mActiveRoute.targetEntity)
+
+        // FIXME: This goes wrong if an npc ever gets stuck or the heights don't match
+        return mActiveRoute.positionsToGo.empty();
+
+      return hasTargetEntityBeenReached(positionNow);
+    }
+
+    bool Pathfinder::isTargetReachedByPosition(const bs::Vector3& position,
+                                               const bs::Vector3& target)
+    {
+      float diffHeight = fabs(position.y - target.y);
+      float diffSide   = bs::Vector2(position.x - target.x, position.z - target.z).length();
+
+      return diffSide < MAX_SIDE_DIFFERENCE_TO_REACH_POSITION &&
+             diffHeight < MAX_HEIGHT_DIFFERENCE_TO_REACH_POSITION;
+    }
+
+    Pathfinder::Instruction Pathfinder::updateToNextInstructionToTarget(
+        const bs::Vector3& positionNow)
+    {
+      Instruction inst;
+
+      if (hasNextRouteTargetBeenReached(positionNow))
+      {
+        if (!mActiveRoute.positionsToGo.empty())
+        {
+          mActiveRoute.positionsToGo.pop_front();
+        }
+      }
+
+      if (hasActiveRouteBeenCompleted(positionNow))
+      {
+        inst.targetPosition = positionNow;
+        return inst;
+      }
+
+      if (shouldReRoute(positionNow))
+      {
+        if (isTargetAnEntity())
+        {
+          startNewRouteTo(positionNow, mActiveRoute.targetEntity);
+        }
+        else
+        {
+          assert(!mActiveRoute.positionsToGo.empty());
+          startNewRouteTo(positionNow, mActiveRoute.positionsToGo.back());
+        }
+      }
+
+      inst.targetPosition = getCurrentTargetPosition(positionNow);
+
+      return inst;
+    }
+
+    bool Pathfinder::hasNextRouteTargetBeenReached(const bs::Vector3& positionNow) const
+    {
+      if (mActiveRoute.positionsToGo.empty())
+      {
+        return isTargetAnEntity() && hasTargetEntityBeenReached(positionNow);
+      }
+
+      return isTargetReachedByPosition(positionNow, mActiveRoute.positionsToGo.front());
+    }
+
+    bool Pathfinder::hasTargetEntityBeenReached(const bs::Vector3& positionNow) const
+    {
+      if (!isTargetAnEntity()) return false;
+
+      bs::Vector3 targetEntityPosition = getTargetEntityPosition();
+
+      return isTargetReachedByPosition(positionNow, targetEntityPosition);
+    }
+
+    bs::Vector3 Pathfinder::getTargetEntityPosition() const
+    {
+      if (!isTargetAnEntity()) return bs::Vector3(0, 0, 0);
+
+      return mActiveRoute.targetEntity->getTransform().pos();
+    }
+
+    bool Pathfinder::isTargetAnEntity() const
+    {
+      return !!mActiveRoute.targetEntity;
```

Comment:
Double negation?

Reply:
That's a way to convert something to a boolean-type. I'll change it to something more expected.

Reply:
Ah I thought `targetEntity` is already boolean. Usually `return mActiveRoute.targetEntity` should work aswell. I can imagine the bsf handles overload `operator bool` similar to standard smart pointers.

Reply:
It's not boolean, but there is an operator converting it to bool. I just wanted to make that more explicit.